ccstools/man/man8/ccs-auditd.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.36.
.TH CCS-AUDITD "8" "May 2009" "ccs-auditd 1.6.8" "System Administration Utilities"
.SH NAME
ccs-auditd \- TOMOYO Linux's auditing daemon
.SH SYNOPSIS
.B ccs-auditd
\fIlocation_to_save_grant_log location_to_save_reject_log\fR
.SH DESCRIPTION
This program reads access request logs from kernel and writes to specified location.
.PP
By running this program upon startup, you can save access logs which violated domain_policy (reject_log) and access logs which didn't violate domain_policy (grant_log) in domain_policy file's format.
.PP
You may specify /dev/null as location to save logs. But in that case, you should set MAX_GRANT_LOG=0 and/or MAX_REJECT_LOG=0 in profile configuration ( /etc/ccs/profile.conf or /proc/ccs/profile ).
.SH EXAMPLES

# ccs\-auditd /dev/null /var/log/tomoyo/reject_log.txt
.SH NOTES

 Start this program from appropriate stage such as /etc/rc.local .
.SH AUTHORS

 penguin-kernel _at_ I-love.SAKURA.ne.jp
.SH COPYRIGHT
Copyright \(co 2005-2009 NTT DATA CORPORATION.
.PP
This program is free software; you may redistribute it under the terms of
the GNU General Public License. This program has absolutely no warranty.


ccstools/ChangeLog
Version 1.0   2005/11/11   First release.

Version 1.0.1 2005/12/08   Minor update release.

  makesyaoranconf.exe:
    Use versionsort() for sorting entries.

  poled.exe:
    Support "search" command.

  poled_old.exe:
    Support "search" command.
    Fix "initializer" checking bug.

  syspol.exe:
    Support editing without resetting cursor position.

Version 1.0.2 2006/02/14   Procedure review.

  savepolicy:
    Support saving "system policy" and "exception policy"
    in addition to "domain policy".

  The following programs were added.

    editpolicy:
      "syspol.exe" "poled.exe" "poled_old.exe" were integrated
      and renamed to "editpolicy".
      This program can edit "system policy", "exception policy"
      and "domain policy".
      Command key assignments were changed.

    checkpolicy:
      A policy validator taken from "poled_old.exe".
      This program was designed for detecting and fixing errors
      in "domain policy".

    loadpolicy:
      A policy reloader.
      This program was designed for loading policy from the disk
      after clearing current policy in the kernel.

    sortpolicy:
      A "domain policy" sorter.
      This program was designed to compress access logs
      generated by "ccs-auditd".
      You can use normal "sort" command for sorting
      "system policy" and "exception policy".

    make_exception.sh:
      A script to create "exception policy".

  The following programs were renamed.

    "remount.exe" was renamed to "remount_rootfs".
    "makesyaoranconf.exe" was renamed to "makesyaoranconf".

  The following programs were removed.

    "poled.exe" "poled_old.exe" "syspol.exe"
    "obsolete_chksymlink" "obsolete_chroot_su"
    "obsolete_lsdir" "obsolete_makelink" "obsolete_movlog"
    "bindtest" "logtest"  "pathnametest" "rofstest"
    "linuxrc_old"

  The following programs for testing TOMOYO Linux's kernel were added.
  They are in the kernel_test directory.

    "sakura_bind_test" "sakura_capability_test"
    "sakura_filesystem_test" "sakura_trace_test"
    "tomoyo_capability_test" "tomoyo_file_test" "tomoyo_info_test"
    "tomoyo_name_test" "tomoyo_port_test" "tomoyo_signal_test"

Version 1.1   2006/04/01   Functionality enhancement release.

  loadpolicy:
    Delete domain for loadpolicy anyway.

  findtemp:
    Now supports for detecting all nonexistent pathnames.

  savepolicy:
    Run twice inside savepolicy itself to include necessary ACLs
    for savepolicy itself.

  The following program for testing TOMOYO Linux's kernel was added.

    "testall.sh"

Version 1.1.1 2006/05/15   Functionality enhancement release.

  The following programs were added.

    ld-watch:
      Monitors /etc/ld.so.cache and updates exception policy.
      This program is used only when updating packages.

    ccs-queryd:
      Monitors /proc/ccs/policy/query for policy violation and
      asks the administrator whether to grant or reject the request.
      This program is used while and after updating packages.

Version 1.1.2 2006/06/02   Functionality enhancement release.

  The following programs were redesigned.

    editpolicy:
      Simplified domain policy handling and removed "save" key.
      All modifications are taken effect immediately.

    loadpolicy:
      Simplified domain policy handling.

    sortpolicy:
      Simplified domain policy handling.

    savepolicy:
      Save all policies by default.

  The following program was removed.

    editpolicy_offline

Version 1.1.3 2006/07/13   Functionality enhancement release.

  The following bugs were fixed.

    editpolicy:
      The "Commands =" line was too wide to show within 80x25 screen.

    checkpolicy:
      Renamed domain for "initializer" was wrong.

Version 1.2   2006/09/03   Functionality enhancement release.

  findtemp:
    Now displays all nonexistent pathnames.

  editpolicy_offline:
    Redesigned to use the same operation manner.
    Saves changes automatically when exiting.

Version 1.3   2006/11/11   First anniversary release.

  The following program was redesigned.

    checkpolicy:
      A policy validator.
      Reads policy from stdin and prints syntax errors with line numbers.

  The following programs were added.

    setprofile:
      Assigns profiles to domains.

    pathmatch:
      Reads pathname patterns and expands them.

    domainmatch:
      fgrep for /proc/ccs/policy/domain_policy .

    ccstree:
      pstree with profile numbers and domain names.

    patternize:
      Reads domain policy and patternize pathnames.

    proxy:
      A tiny TCP port forwarder, binding to local port explicitly
      to allow servers filter based on client's port numbers.

    mailauth:
      An example program for CERBERUS.

    timeauth:
      An example program for CERBERUS, similar to honey.

  The following programs were removed.
  If you need them, please take from version 1.2 .
    "remount_rootfs" "linuxrc"
    "dumplink" "dumpsymlink" "makelink" "makesymlink"

  The following program for testing TOMOYO Linux's kernel was added.

    "tomoyo_rewrite_test"

Version 1.3.1 2006/12/08   Minor update release.

  The following bug was fixed.

    editpolicy:
      PageUp/PageDown keys and screen drawings were not working well
      on some environments due to forcefully setting "TERM=linux".

  The following program for testing TOMOYO Linux's kernel was updated.

    "newns"

Version 1.3.2 2007/02/14   Usability enhancement release.

  Many tools were merged into single source code.
  Policy editor was redesigned.

Version 1.4   2007/04/01   x86_64 support release.

  The following bug was fixed.

    editpolicy:
      Domain flags was wrong if "keep_domain <kernel>" is given.

Version 1.4.1 2007/06/05   Minor update release.

  Single source code was divided into many source code.

  The following bug was fixed.

    checkpolicy:
      "keep_domain" syntax was not checked correctly.

Version 1.4.2 2007/07/13   Bug fix release.

  .init:

ccstools/src/ccstools.src/ccs-auditd.c
/*
 * ccs-auditd.c
 *
 * TOMOYO Linux's utilities.
 *
 * Copyright (C) 2005-2009  NTT DATA CORPORATION
 *
 * Version: 1.6.8p1   2009/06/23
 *
 */
#include "ccstools.h"

int ccsauditd_main(int argc, char *argv[])
{
	const char *procfile_path[CCS_AUDITD_MAX_FILES] = {
		proc_policy_grant_log,
		proc_policy_reject_log
	};
	int i;
	int fd_in[CCS_AUDITD_MAX_FILES];
	int fd_out[CCS_AUDITD_MAX_FILES];
	const char *logfile_path[2] = { "/dev/null", "/dev/null" };
	if (access(procfile_path[0], R_OK) || access(procfile_path[1], R_OK)) {
		fprintf(stderr, "You can't run this daemon for this kernel.\n");
		return 0;
	}
	if (argc < 3) {
		fprintf(stderr, "%s grant_log_file reject_log_file\n"
			"  These files may /dev/null, if needn't to be saved."
			"\n", argv[0]);
		return 0;
	}
	logfile_path[0] = argv[1];
	logfile_path[1] = argv[2];
	{ /* Get exclusive lock. */
		int fd = open("/proc/self/exe", O_RDONLY);
		if (flock(fd, LOCK_EX | LOCK_NB) == EOF)
			return 0;
	}
	umask(0);
	for (i = 0; i < CCS_AUDITD_MAX_FILES; i++) {
		fd_out[i] = open(logfile_path[i], O_WRONLY | O_CREAT |
				 O_APPEND, 0600);
		if (fd_out[i] == EOF) {
			fprintf(stderr, "Can't open %s for writing.\n",
				logfile_path[i]);
			return 1;
		}
	}
	switch (fork()) {
	case 0:
		break;
	case -1:
		fprintf(stderr, "Can't fork()\n");
		return 1;
	default:
		return 0;
	}
	if (setsid() == EOF) {
		fprintf(stderr, "Can't setsid()\n");
		return 1;
	}
	switch (fork()) {
	case 0:
		break;
	case -1:
		fprintf(stderr, "Can't fork()\n");
		return 1;
	default:
		return 0;
	}
	if (chdir("/")) {
		fprintf(stderr, "Can't chdir()\n");
		return 1;
	}
	close(0);
	close(1);
	close(2);
	openlog("ccs-auditd", 0,  LOG_USER);
	for (i = 0; i < CCS_AUDITD_MAX_FILES; i++) {
		fd_in[i] = open(procfile_path[i], O_RDONLY);
		if (fd_in[i] == EOF) {
			syslog(LOG_WARNING, "Can't open %s for reading.\n",
			       procfile_path[i]);
			return 1;
		}
	}
	syslog(LOG_WARNING, "Started.\n");
	while (true) {
		char buffer[16384];
		char timestamp[128];
		fd_set rfds;
		FD_ZERO(&rfds);
		for (i = 0; i < CCS_AUDITD_MAX_FILES; i++)
			FD_SET(fd_in[i], &rfds);
		/* Wait for data. */
		if (select(FD_SETSIZE, &rfds, NULL, NULL, NULL) == EOF)
			break;
		for (i = 0; i < CCS_AUDITD_MAX_FILES; i++) {
			time_t stamp;
			char *cp;
			int len;
			struct tm *tm;
			if (!FD_ISSET(fd_in[i], &rfds))
				continue;
			memset(buffer, 0, sizeof(buffer));
			if (read(fd_in[i], buffer, sizeof(buffer) - 1) < 0)
				continue;
			memset(timestamp, 0, sizeof(timestamp));
			if (sscanf(buffer, "#timestamp=%lu", &stamp) != 1)
				goto no_timestamp;
			cp = strchr(buffer, ' ');
			if (!cp)
				goto no_timestamp;
			tm = localtime(&stamp);
			snprintf(timestamp, sizeof(timestamp) - 1,
				 "#%04d-%02d-%02d %02d:%02d:%02d#",
				 tm->tm_year + 1900, tm->tm_mon + 1,
				 tm->tm_mday, tm->tm_hour, tm->tm_min,
				 tm->tm_sec);
			memmove(buffer, cp, strlen(cp) + 1);
no_timestamp:
			/* Open destination file. */
			if (access(logfile_path[i], F_OK)) {
				close(fd_out[i]);
				fd_out[i] = open(logfile_path[i],
						 O_WRONLY | O_CREAT | O_APPEND,
						 0600);
				if (fd_out[i] == EOF) {
					syslog(LOG_WARNING,
					       "Can't open %s for writing.\n",
					       logfile_path[i]);
					goto out;
				}
			}
			len = strlen(timestamp);
			write(fd_out[i], timestamp, len);
			len = strlen(buffer);
			write(fd_out[i], buffer, len);
			write(fd_out[i], "\n", 1);
			fsync(fd_out[i]);
		}
	}
out:
	syslog(LOG_WARNING, "Terminated.\n");
	closelog();
	return 0;
}
